README.md
JC2MP Deathmatch
================
Author: m1nd0 (joopselen@gmail.com)

Description
------------
Deathmatch mode for JC2MP. 


Installation
------------
Drop in /scripts/ folder.
Edit /server/DeathMatchManager.lua with your own admin ID.
Edit /server/arenas/Manifest.txt for desired arena's


How it works
------------
**Client**
- pres 'k' or type /dm or /deathmatch to open the GUI

**Admin**
- Start: starts the event (disregarding minplayers, and any timers)
- Debugstart: starts the event (disregarding minplayers, and any timers) and doesn't check finish criteria for first 60 seconds

ARENA SETTINGS
--------------
- Location(Name): Name
- Players(min,max): Min players before match can start. Max players must equal spawn points
- Boundary(x, y, z, radius): Center point of arena. Radius defines radius, if player leaves this he will be killed within 20 seconds. 
- MaximumY(ymax): maximum y
- MinimumY(ymin): minimum y
- GrapplingHookAllowed(bool): whether the grapplinghook is allowed to be used
- ParachuteAllowed(bool): whether the parachute is allowed to be used
- Spawn(x, y, z, dir, 0, 0): A spawn location. Multiple fore more locations, must match maxplayers.
- Weapon(Handgun): Weapons pool, 1 weapon will be randomly drawn from all weapons. New line for each weapon. 
	Examples of weapons:
	Handgun,Revolver,SMG,SawnOffShotgun,Assault,Shotgun,Sniper,MachineGun
	Or Weapon(id,ammo,spare ammo)

Credits
-------
Patawic: I used his derby script as a basis for this script.

changelog.md
## 0.2.1 (2013-12-23)
Features:
  
  - joinall button for admins

Bugfixes:

  - fixed area size of hotels arena
  - fixed error where leave button wouldn't switch back to join after match ended

## 0.2.0 (2013-12-23)

Features:

  - GUI for joining deathmatches (binded to keyboard 'k', chat /dm and /deathmatch)
  - all available deathmatches are automatically created (not random any more)
  - admins can be defined in admins.txt
  - admins have more options in GUI (start & debugstart)
  - players are removed from queue if they leave the default world
  - increased default ammo to 999
  
Bugfixes:

  - fix for linux casesensitive filenames with arenas
  - fix for players falling though ground at spawn

server/DeathMatch.lua
function math.round(x)
	if x%2 ~= 0.5 then
		return math.floor(x+0.5)
	end
	return x-0.5
end
class "DeathMatch"
function DeathMatch:__init(manager, world, arenaName)
	self.deathMatchManager = manager
	self.world = world
	
	self.arenamanager = Arena(self.deathMatchManager)
	self.arena = self.arenamanager:LoadArena(arenaName)

	self:InitVars()
	
	Events:Subscribe("PostTick", self, self.PostTick)
	Events:Subscribe("JoinGamemode", self, self.JoinGamemode)
	Events:Subscribe("PlayerDeath", self, self.PlayerDeath)
	Events:Subscribe("PlayerQuit", self, self.PlayerLeave)
	Events:Subscribe("ModuleUnload", self, self.ModuleUnload)
end

---------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------EVENTS----------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function DeathMatch:PostTick()
	if (self.state == "Lobby") then
		if ((self.numPlayers >= self.minPlayers and self.startTimer:GetSeconds() > 30) or (self.numPlayers >= self.minPlayers and self.globalStartTimer:GetSeconds() > 300)) then
			self:Start()
		end
	elseif (self.state == "Setup") then
		if (self.setupTimer:GetSeconds() > 10) then
			self.countdownTimer = Timer()
			--set state
			self.state = "Countdown"
			self:SetClientState()
			self.setupTimer = nil
		end
	elseif (self.state == "Countdown") then
		if (self.countdownTimer:GetSeconds() > 10) then
			--set state
			self.state = "Running"
			self:SetClientState()
			self.countdownTimer = nil
			self.deathMatchTimer = Timer()
		end
		self:CheckBoundaries()
	elseif (self.state == "Running") then
		--Check if players need to be prepped
		if(self.playerPrepped == false) then
			self:PreparePlayers()
			self.playerPrepped = true
		end
		
		--player gets killed when out of boundries
		self:CheckBoundaries()
			
		--Actively check for players & handle DeathMatch ending (if debug 60 secs later cause of test)
		if (self.debugMode == false or self.deathMatchTimer:GetSeconds() > 60) then
			self:CheckPlayers()
		end
	end
end

function DeathMatch:PlayerDeath(args)
	if self:HasPlayer(args.player) then
		if (self.state ~= "Lobby" and args.player:GetWorld() == self.world) then
			local numberEnding = ""
			local lastDigit = self.numPlayers % 10
			if ((self.numPlayers < 10) or (self.numPlayers > 20 and self.numPlayers < 110) or (self.numPlayers > 120)) then
				if (lastDigit  == 1) then
					numberEnding = "st"
				elseif (lastDigit == 2) then
					numberEnding = "nd"
				elseif (lastDigit == 3) then
					numberEnding = "rd"
				else
					numberEnding = "th"
				end
			else
				numberEnding = "th"
			end
			self.deathMatchManager:MessagePlayer(args.player, "Congratulations you came " ..tostring(self.numPlayers) .. numberEnding)
			self:RemovePlayer(args.player)

			local currentMoney = args.player:GetMoney()
			local addMoney = math.ceil(100 * math.exp(self.scaleFactor * (self.startPlayers - self.numPlayers))) / 2
			args.player:SetMoney(currentMoney + addMoney)
		end
	end
end

function DeathMatch:PlayerLeave(args)
	if (self:HasPlayer(args.player)) then
		self:RemovePlayer(args.player)
	end
end

function DeathMatch:SetClientState(newstate)
	for index,player in pairs(self.players) do
		if newstate == nil then
			Network:Send(player, "SetState", self.state)
		else
			Network:Send(player, "SetState", newstate)
		end
	end
end

function DeathMatch:SendSettingsToClient()
	local settings = {}
	settings.grapplingAllowed = self.grapplingAllowed
	settings.parachuteAllowed = self.parachuteAllowed

	for index,player in pairs(self.players) do
		Network:Send(player, "SetAllowedItems", settings)
	end
end

function DeathMatch:UpdatePlayerCount()  
	for id ,player in pairs(self.players) do
		Network:Send(player, "PlayerCount", self.numPlayers)
	end
end

function DeathMatch:ModuleUnload()
	for k,p in pairs(self.eventPlayers) do
		if (self.state ~= "Lobby") then
			p:Leave()
			self.deathMatchManager:MessagePlayer(p.player, "DeathMatch script unloaded. You have been restored to your starting position.")
			self:SetClientState("Inactive")
		end
	end
end
function DeathMatch:JoinGamemode( args )
	if args.name ~= "DeathMatch" then
		self:RemovePlayer(args.player)
	end
end
---------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function DeathMatch:PreparePlayers()
	for k,p in pairs(self.eventPlayers) do
		p.player:SetHealth(1)
		p.player:GiveWeapon(self.weapon.slot, Weapon( self.weapon.id, self.weapon.ammo1, self.weapon.ammo2))
	end
end
function DeathMatch:CheckBoundaries()
	for k,p in pairs(self.players) do
		local boundary = self.arena.Boundary.position
		local radius = self.arena.Boundary.radius
		local distance = (p:GetPosition():Distance2D(boundary)) 
		
		--CHECK IS PLAYER IS OUTSIDE THE EVENT BOUNDARIES
		if(p:GetWorld() == self.world) then
			if (distance > radius or p:GetPosition().y > self.arena.MaximumY or p:GetPosition().y < self.arena.MinimumY) then
				if (p.timer ~= nil) then
					if p.timer:GetSeconds() > 20 then
						--Kill the player
						p:SetHealth(0)
						p.timer = nil
					end
				else
					p.timer = Timer()
					p.outOfArena = true
					Network:Send(p, "OutOfArena")
				end
			elseif (p.outOfArena) then
				p.outOfArena = false
				p.timer = nil
				Network:Send(p, "BackInArena")
			end
		end
	end
end
function DeathMatch:CheckPlayers()
	if (self.numPlayers == 1 and self.state ~= "Lobby") then
	--kick everyone out and broadcast the winner
		for k,p in pairs(self.players) do
			self.deathMatchManager:MessageGlobal(p:GetName() .. " has won the Deathmatch event " .. self.arena.Location .. "!")

			local currentMoney = p:GetMoney()
			local addMoney = math.ceil(100 * math.exp(self.scaleFactor * (self.startPlayers - self.numPlayers))) / 2
			p:SetMoney(currentMoney + addMoney)
			self:RemovePlayer(p, "Congratulations you came 1st!")
		end
		self:Cleanup()
	elseif (self.numPlayers == 0) then
		self:Cleanup()
	end
end
---------------------------------------------------------------------------------------------------------------------
--------------------------------------------------EVENT START--------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function DeathMatch:Start()
	self.state = "Setup"
	self.startPlayers = self.numPlayers
	self.setupTimer = Timer()
	self:SendSettingsToClient()
	self:SetClientState()
	
	local tempPlayers = {}
	for id , player in pairs(self.players) do
		table.insert(tempPlayers , player)
	end
	local divider = math.floor(self.maxPlayers / self.numPlayers)
	local idInc = 1

	for index, player in ipairs(tempPlayers)do 
		if (player:GetHealth() == 0) then
			self:RemovePlayer(player, "You have been removed from the DeathMatch event.")
		else
			self:SpawnPlayer(player, tonumber(math.round(idInc)))
		end
		idInc = idInc + divider
	end

	self.highestMoney = self.startPlayers * 400
	self.scaleFactor = math.log(self.highestMoney/100)/self.startPlayers
end

function DeathMatch:SpawnPlayer(player, index)
	if (IsValid(self.arena.SpawnPoint[index]) ~= nil) then
		--TELEPORT THE PLAYER
		player:SetWorld(self.world)
		player:SetPosition(self.arena.SpawnPoint[index].position + Vector3(0,1,0))
		player:SetAngle(self.arena.SpawnPoint[index].angle)
		player:ClearInventory()

		local p = self.eventPlayers[player:GetId()]
		p.deathMatchPosition = self.arena.SpawnPoint[index].position
		p.deathMatchAngle = self.arena.SpawnPoint[index].angle
		
	else
		self:RemovePlayer(player, "An error occured, you were removed from the DeathMatch.")
	end

end
---------------------------------------------------------------------------------------------------------------------
-------------------------------------------PLAYER JOINING/LEAVING----------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function DeathMatch:HasPlayer(player)
	return self.players[player:GetId()] ~= nil
end

function DeathMatch:JoinPlayer(player)
	if (player:GetWorld() ~= DefaultWorld) then
		self.deathMatchManager:MessagePlayer(player, "You must exit other gamemodes before you can join.")
	else
		if (self.state == "Lobby") then
			local p = Player(player)
			self.eventPlayers[player:GetId()] = p
			self.players[player:GetId()] = player

			self.deathMatchManager.playerIds[player:GetId()] = true
			self.numPlayers = self.numPlayers + 1
			--self:MessagePlayer(player, "You have been entered into the next DeathMatch event! It will begin shortly.") 

			Network:Send(player, "SetState", "Lobby")
			self:UpdatePlayerCount()
			self.startTimer:Restart()
			
			if(self.numPlayers == 1) then
				 self.globalStartTimer:Restart()
			end
			if (self.numPlayers == self.maxPlayers) then
				self:Start()
			end
		end
	end
end

function DeathMatch:RemovePlayer(player, message)
	if message ~= nil then
		self.deathMatchManager:MessagePlayer(player, message)    
	end
	
	local p = self.eventPlayers[player:GetId()]
	if p == nil then 
		return nil
	else
		self.players[player:GetId()] = nil
		self.eventPlayers[player:GetId()] = nil
		self.deathMatchManager.playerIds[player:GetId()] = nil
		self.numPlayers = self.numPlayers - 1
		if (self.state ~= "Lobby") then
			p:Leave()
		end
		Network:Send(player, "SetState", "Inactive")
		Network:Send(player, "ArenaName", nil)
		self:UpdatePlayerCount()
	end
end
---------------------------------------------------------------------------------------------------------------------
---------------------------------------------------CLEANUP-----------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function DeathMatch:Cleanup()
	self.state = "Cleanup"
	for index, player in pairs(self.players) do
		self:RemovePlayer(player)
	end
	self.state = "Lobby"
end

function DeathMatch:InitVars()
	self.weapon = table.randomvalue(self.arena.Weapons)
	self.debugMode = false
	self.playerPrepped = false
	self.state = "Lobby"
	self.startTimer = Timer()
	self.players = {}
	self.eventPlayers = {}
	self.grapplingAllowed = self.arena.grapplingAllowed
	self.parachuteAllowed = self.arena.parachuteAllowed
	self.minPlayers = self.arena.minPlayers
	self.maxPlayers = self.arena.maxPlayers
	self.startPlayers = 0
	self.numPlayers = 0
	self.highestMoney = 0
	self.scaleFactor = 0

	self.globalStartTimer = Timer()
	self.setupTimer = nil
	self.countdownTimer = nil
	self.deathMatchTimer = nil
end

server/DeathMatchManager.lua
class "DeathMatchManager"
function DeathMatchManager:__init()
	self.players = {}
	self.playerIds = {}
	self.admins = {}
	self:LoadAdmins()
	
	self.arenamanager = Arena(self)
	self.deathmatches = {}
	
	self.clientsListening = {}
	--create arena for each dm
	for index, arenaName in pairs(self.arenamanager.arenas) do
		self.deathmatches[arenaName] = DeathMatch(self, World.Create(), arenaName)
	end
	
	self.timerInfo = Timer()
	self.timerInstance = Timer()
   
	Events:Subscribe("PlayerJoin", self, self.PlayerJoin)
	Events:Subscribe("PlayerChat", self, self.ChatMessage)
	Events:Subscribe("PostTick", self, self.PostTick)
	Network:Subscribe("DeathMatchOpenArenaWindow", self, self.OpenArenaWindow)
	Network:Subscribe("DeathMatchJoinArena", self, self.JoinArena)
	Network:Subscribe("DeathMatchJoinAll", self, self.JoinAll)
	Network:Subscribe("DeathMatchStart", self, self.Start)
	
	--If module is reloaded make sure admins are set to admin again
	self.adminCheckTimer = Timer()
end

function DeathMatchManager:LoadAdmins()
	local path = "/admins.txt"
	local tempFile , tempFileError = io.open(path , "r")
	if tempFileError then
			print()
			print("*ERROR*")
			print(tempFileError)
			print()
			fatalError = true
			return
	else
			io.close(tempFile)
	end
	
	for line in io.lines(path) do
		
		if line:sub(1,1) == "A" then
			line = line:gsub("Admin%(", "")
			line = line:gsub("%)", "")
			line = line:gsub(" ", "")
			local tokens = line:split(",") 
			
			self.admins[tokens[2]] = tokens[1]
		end
	end
end


function DeathMatchManager:Start(args, player)
	if (self:IsAdmin(player)) then
		self.deathmatches[args.arenaName].debugMode = args.debugMode
		self.deathmatches[args.arenaName]:Start()
		local deathmatchInfo = self:GenereateDMInfo()
		self:SendDeathmatchInfoToClient(player, deathmatchInfo)
	end
end

function DeathMatchManager:IsAdmin(player)
	return self.admins[tostring(player:GetSteamId())] ~= nil
end

function DeathMatchManager:JoinAll(args, player)
	if (self:IsAdmin(player)) then
		local isInAny = false
		for p in Server:GetPlayers() do 
			for arenaName, deathmatch in pairs(self.deathmatches) do
				if(deathmatch:HasPlayer(p) == true) then
					isInAny = true
				end
			end
			if(isInAny == false) then
				self.deathmatches[args.arenaName]:JoinPlayer(p)
			end
		end
	end
	local deathmatchInfo = self:GenereateDMInfo()
	self:SendDeathmatchInfoToClient(player, deathmatchInfo)
end

function DeathMatchManager:PlayerJoin(args)
	local player = args.player
	
	if (self:IsAdmin(player)) then
		Network:Send(player, "SetIsAdmin", true)
	end
end

function DeathMatchManager:JoinArena(args, player)
	--Remove from other dm if needed
	local joinAgain = true
	for arenaName, deathmatch in pairs(self.deathmatches) do
		if(deathmatch:HasPlayer(player) == true) then
			deathmatch:RemovePlayer(player)
			if(args.arenaName == arenaName) then
				joinAgain = false
			end
		end
	end
	
	local arenaName = nil
	if(joinAgain) then
		self.deathmatches[args.arenaName]:JoinPlayer(player)
		arenaName = args.arenaName
	end
	
	--update arenaName on client
	Network:Send(player, "ArenaName", arenaName)
	local deathmatchInfo = self:GenereateDMInfo()
	self:SendDeathmatchInfoToClient(player, deathmatchInfo)
end

--Clients opens window, from now on start sending update messages
function DeathMatchManager:OpenArenaWindow(open, player)
	if(open) then
		self.clientsListening[player:GetId()] = player
		local deathmatchInfo = self:GenereateDMInfo()
		self:SendDeathmatchInfoToClient(player, deathmatchInfo)
	else
		self.clientsListening[player:GetId()] = nil
	end
end

function DeathMatchManager:PostTick()
	if(self.timerInfo:GetSeconds() >= 5) then
		local deathmatchInfo = self:GenereateDMInfo()
		
		for index, player in pairs(self.clientsListening) do
			self:SendDeathmatchInfoToClient(player, deathmatchInfo)
		end
		self.timerInfo:Restart()
	end
	
	--Resend p admins stats if script is reloaded
	if(self.adminCheckTimer ~= nil) then
		if( self.adminCheckTimer:GetSeconds() >= 5) then
			for p in Server:GetPlayers() do 
				self:PlayerJoin({player = p}) 
			end
			self.adminCheckTimer = nil
		end
	end
	
	if(self.timerInstance:GetSeconds() > 2) then
		for arenaName, deathmatch in pairs(self.deathmatches) do
			if(deathmatch.state == "Lobby") then
				for index, player in pairs(deathmatch.players) do
					if(player:GetWorld() ~= DefaultWorld) then
						deathmatch:RemovePlayer(player, "You are no longer in the default instance and are removed from the deathmatch queue.")
					end
				end
			end
		end
		self.timerInstance:Restart()
	end
end

function DeathMatchManager:SendDeathmatchInfoToClient(player, deathmatchInfo)
	Network:Send(player, "DeathmatchInfo", deathmatchInfo)
end

function DeathMatchManager:GenereateDMInfo()
	local deathmatchInfo = {}
	for index, deathmatch in pairs(self.deathmatches) do
		local info = {}
		info.location = deathmatch.arena.Location
		info.name = deathmatch.arena.name
		info.state = deathmatch.state
		info.curPlayers = deathmatch.numPlayers
		info.minPlayers = deathmatch.arena.minPlayers
		info.maxPlayers = deathmatch.arena.maxPlayers
		info.grapplingAllowed = deathmatch.arena.grapplingAllowed
		info.parachuteAllowed = deathmatch.arena.parachuteAllowed
		table.insert(deathmatchInfo, info)
	end
	return deathmatchInfo
end

-------------
--CHAT STUF--
-------------
function DeathMatchManager:MessagePlayer(player, message)
	player:SendChatMessage( "[DeathMatch] " .. message, Color(30, 200, 220) )
end

function DeathMatchManager:MessageGlobal(message)
	Chat:Broadcast( "[DeathMatch] " .. message, Color(0, 255, 255) )
end

function DeathMatchManager:ChatMessage(args)
	local msg = args.text
	local player = args.player
	
	-- If the string is't a command, we're not interested!
	if ( msg:sub(1, 1) ~= "/" ) then
		return true
	end    
	
	local cmdargs = {}
	for word in string.gmatch(msg, "[^%s]+") do
		table.insert(cmdargs, word)
	end
	
	if (self:IsAdmin(player)) then
		--admins can output points to file for easy creation of arenas spawns 
		if (cmdargs[1] == "/p") then
			local text = "Spawn(" .. tostring(player:GetPosition()) .. "," .. tostring(player:GetAngle()) .. ")\n"
			local file = io.open("server/Arenas/SpawnPointsOutput.txt" , "a+")
			file:write(text)
			file:close()
		end
		
		--if (cmdargs[1] == "/port") then
		--	player:SetPosition(Vector3(14014.245117, 458.688324, 14270.292969))
		--end
	end
	return false
end

server/Arena.lua
class "Arena"
function Arena:__init(manager)
	self.deathMatchManager = manager
	self.arenaRootpath = "/server/Arenas/"
	self.manifestPath = self.arenaRootpath .. "Manifest.txt"
	--TODO: Fix this ugly stuff
	--Arena names
	self.arenaNames = {}
	self.numArenas = 0
	
	--ClassNames
	self.arenas = {}
	

	
	self.ammo_counts            = {
	[2] = { 12, 999 }, [4] = { 7, 999 }, [5] = { 30, 999 },
	[6] = { 3, 999 }, [11] = { 20, 999 }, [13] = { 6, 999 },
	[14] = { 4, 999 }, [16] = { 3, 999 }, [17] = { 5, 999 },
	[28] = { 26, 999 } 
	}
	
	self.settings = {}
	
	self:LoadManifest(self.manifestPath)
	self:DetermineClass()
end
---------------------------------------------------------------------------------------------------------------------
------------------------------------------------MANIFEST LOADING-----------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function Arena:LoadManifest(path)
	local tempFile , tempFileError = io.open(path , "r")
	if tempFileError then
			print()
			print("*ERROR*")
			print(tempFileError)
			print()
			fatalError = true
			return
	else
			io.close(tempFile)
	end
	-- Loop through each line in the manifest.
	for line in io.lines(path) do
		-- Make sure this line has stuff in it.
		if string.find(line , "%S") then
				-- Add the entire line, sans comments, to self.arenaNames
				table.insert(self.arenaNames , line)
				self.numArenas = self.numArenas + 1
		end
	end
end
---------------------------------------------------------------------------------------------------------------------
------------------------------------------------DETERMINE CLASSES----------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function Arena:DetermineClass()
	for index, arena in pairs(self.arenaNames) do
		local path = self.arenaRootpath .. arena .. ".arena"
		--check if path is invalid
		if path == nil then
			print("*ERROR* - Arena path is nil!")
			return nil
		end	
		local file = io.open(path , "r") 
		--check if file exists
		if not file then
			print("*ERROR* - Cannot open arena file: "..path)
			return nil
		end
		
		table.insert(self.arenas, arena)
	end
end
---------------------------------------------------------------------------------------------------------------------
-----------------------------------------------ARENA FILE PARSING---------------------------------------------------
---------------------------------------------------------------------------------------------------------------------
function Arena:LoadArena(name)
	if name == nil then
		name = self:PickArena()
	end
	local path = self.arenaRootpath .. name .. ".arena"
	--check if path is invalid
	if path == nil then
		print("*ERROR* - Arena path is nil!")
		return nil
	end	
	local file = io.open(path , "r") 
	--check if file exists
	if not file then
		print("*ERROR* - Cannot open Arena file: "..path)
		return nil
	end

	local arena = {}
	arena.name = name
	arena.Location = nil
	arena.minPlayers = nil
	arena.maxPlayers = nil
	arena.SpawnPoint = {}
	arena.Boundary = {}
	arena.MinimumY = nil
	arena.MaximumY = nil
	arena.Weapons = {}
	arena.grapplingAllowed = true
	arena.parachuteAllowed = true
	
	--loop through file line by line
	for line in file:lines() do
		if line:sub(1,1) == "L" then
			arena.Location =  self:Location(line)
		elseif line:sub(1,2) == "Pl" then
			local playerCount = self:Players(line)
			arena.minPlayers = playerCount.minPlayers
			arena.maxPlayers = playerCount.maxPlayers
		elseif line:sub(1,1) == "B" then
			local boundary = self:Boundary(line)
			arena.Boundary.position = boundary.position
			arena.Boundary.radius = boundary.radius
		elseif line:sub(1,1) == "M" and line:sub(2,2) == "i"then
			arena.MinimumY = self:MinimumY(line)
		elseif line:sub(1,1) == "M" and line:sub(2,2) == "a"then
			arena.MaximumY = self:MaximumY(line)
		elseif line:sub(1,1) == "S" then
			table.insert(arena.SpawnPoint, self:Spawn(line))
		elseif line:sub(1,1) == "W" then
			table.insert(arena.Weapons, self:Weapon(line))
		elseif line:sub(1,1) == "G" then
			arena.grapplingAllowed = self:GetGrapplingAllowed(line)
		elseif line:sub(1,2) == "Pa" then
			arena.parachuteAllowed = self:GetParachuteAllowed(line)
		end
	end
	return arena
end
function Arena:Location(line)
	line = line:gsub("Location%(", "")
	line = line:gsub("%)", "")

	return line
end
function Arena:GetGrapplingAllowed(line)
	line = line:gsub("GrapplingHookAllowed%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")
	
	if(line == "true") then
		return true
	else
		return false
	end
end
function Arena:GetParachuteAllowed(line)
	line = line:gsub("ParachuteAllowed%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")
	
	if(line == "true") then
		return true
	else
		return false
	end
end
function Arena:Weapon(line)
	line = line:gsub("Weapon%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")
	local tokens = line:split(",") 
	
	weapon = {}
	
	if(tokens[1] == "Handgun") then
		weapon.id = Weapon.Handgun
		weapon.slot = WeaponSlot.Right
	elseif(tokens[1] == "Revolver") then
		weapon.id = Weapon.Revolver
		weapon.slot = WeaponSlot.Right
	elseif(tokens[1] == "SMG") then
		weapon.id = Weapon.SMG
		weapon.slot = WeaponSlot.Right	
	elseif(tokens[1] == "SawnOffShotgun") then
		weapon.id = Weapon.SawnOffShotgun
		weapon.slot = WeaponSlot.Right
	elseif(tokens[1] == "Assault") then
		weapon.id = Weapon.Assault
		weapon.slot = WeaponSlot.Primary	
	elseif(tokens[1] == "Shotgun") then
		weapon.id = Weapon.Shotgun
		weapon.slot = WeaponSlot.Primary		
	elseif(tokens[1] == "Sniper") then
		weapon.id = Weapon.Sniper
		weapon.slot = WeaponSlot.Primary	
	elseif(tokens[1] == "MachineGun") then
		weapon.id = Weapon.MachineGun
		weapon.slot = WeaponSlot.Primary
	end
	
	--If weapon unknown parse rest of line for id and ammo counts
	if(weapon.id == nil) then
		weapon.id = tokens[1]
	end
	
	--load custom ammo counts or pre defined
	if(tokens[2] ~= nil) then
		weapon.ammo1 = tokens[2]
		weapon.ammo2 = tokens[3]
	else
		weapon.ammo1 = self.ammo_counts[weapon.id][1]
		weapon.ammo2 = self.ammo_counts[weapon.id][2]
	end
	
	return weapon
end
function Arena:Players(line)
	line = line:gsub("Players%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")

	local tokens = line:split(",")   
	local args = {}

	args.minPlayers = tonumber(tokens[1])
	args.maxPlayers = tonumber(tokens[2])

	return args
end
function Arena:Boundary(line)
	line = line:gsub("Boundary%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")

	local tokens = line:split(",")   
	local args = {}
	-- Create tables containing appropriate strings
	args.position	= Vector3(tonumber(tokens[1]), tonumber(tokens[2]), tonumber(tokens[3]))
	args.radius		= tonumber(tokens[4])

	return args
end
function Arena:MinimumY(line)
	line = line:gsub("MinimumY%(", "")
	line = line:gsub("%)", "")

	return tonumber(line)
end
function Arena:MaximumY(line)
	line = line:gsub("MaximumY%(", "")
	line = line:gsub("%)", "")

	return tonumber(line)
end
function Arena:Spawn(line)
	line = line:gsub("Spawn%(", "")
	line = line:gsub("%)", "")
	line = line:gsub(" ", "")

	local tokens = line:split(",")   
	local args = {}
	-- Create tables containing appropriate strings
	args.position	= Vector3(tonumber(tokens[1]), tonumber(tokens[2]), tonumber(tokens[3]))
	args.angle		= Angle(tonumber(tokens[4]), tonumber(tokens[5]), tonumber(tokens[6]))

	return args
end
function Arena:PickArena()
	return table.randomvalue(self.arenas)
end